circuits/src/zk_gadgets/poseidon.rs

```diff
@@ -142,451 +127,217 @@ impl PoseidonHashGadget {
 
     /// Squeeze an output from the hasher, and constraint its value to equal the
     /// provided statement variable.
-    pub fn constrained_squeeze<L, CS>(&mut self, expected: L, cs: &mut CS) -> Result<(), R1CSError>
-    where
-        L: Into<LinearCombination>,
-        CS: RandomizableConstraintSystem,
-    {
+    pub fn constrained_squeeze<C: ConstraintSystem<ScalarField>>(
+        &mut self,
+        expected: Variable,
+        cs: &mut C,
+    ) -> Result<(), PlonkError> {
         let squeezed_elem = self.squeeze(cs)?;
-        cs.constrain(squeezed_elem - expected);
+        cs.enforce_equal(expected, squeezed_elem)?;
         Ok(())
     }
 
     /// Squeeze a set of elements from the hasher, and constraint the elements
     /// to be equal to the provided statement variables
-    pub fn batch_constrained_squeeze<L, CS>(
+    pub fn batch_constrained_squeeze<C: ConstraintSystem<ScalarField>>(
         &mut self,
-        expected: &[L],
-        cs: &mut CS,
-    ) -> Result<(), R1CSError>
-    where
-        L: LinearCombinationLike,
-        CS: RandomizableConstraintSystem,
-    {
+        expected: &[Variable],
+        cs: &mut C,
+    ) -> Result<(), PlonkError> {
         expected
             .iter()
-            .try_for_each(|val| self.constrained_squeeze(val.clone(), cs))
+            .try_for_each(|val| self.constrained_squeeze(*val, cs))
     }
 
-    /// Permute the digest by applying the Poseidon round function
-    fn permute<CS: RandomizableConstraintSystem>(&mut self, cs: &mut CS) -> Result<(), R1CSError> {
-        // Compute full_rounds / 2 rounds in which the sbox is applied to all elements
-        for round in 0..self.params.full_rounds / 2 {
-            self.add_round_constants(round);
-            self.apply_sbox(true /* full_round */, cs)?;
-            self.apply_mds()?;
+    /// Permute the state using the Poseidon 2 permutation
+    #[allow(clippy::missing_docs_in_private_items)]
```

Comment:
nit: do we still need this lint allowance? there's a doc comment here now

Reply:
Yes, its for the constant in the function